h2o-algos/src/main/java/hex/tree/DTree.java

```diff
@@ -1364,6 +1461,424 @@ static Split findBestSplitPoint(DHistogram hs, int col, double min_rows, int con
     return split;
   }
 
+  static Split findBestSplitPointUplift(DHistogram hs, int col, double min_rows) {
+    if(hs._valsUplift == null) {
+      if (LOG.isTraceEnabled()) LOG.trace("can't split " + hs._name + ": histogram not filled yet.");
+      return null;
+    }
+    final int nbins = hs.nbins();
+    assert nbins >= 1;
+
+    final Divergence upliftMetric = hs._upliftMetric;
+
+    // Histogram arrays used for splitting, these are either the original bins
+    // (for an ordered predictor), or sorted by the mean response (for an
+    // unordered predictor, i.e. categorical predictor).
+    double[]   vals =   hs._vals;
+    final int vals_dim = hs._vals_dim;
+    double[] valsUplift = hs._valsUplift;
+    final int valsUpliftDim = 4;
+    int idxs[] = null;          // and a reverse index mapping
+
+    // For categorical (unordered) predictors, sort the bins by average
+    // prediction then look for an optimal split.
+    if( hs._isInt == 2 && hs._step == 1 ) {
+      // Sort the index by average response
+      idxs = MemoryManager.malloc4(nbins+1); // Reverse index
+      for( int i=0; i<nbins+1; i++ ) idxs[i] = i; //index in 0..nbins-1
+      final double[] avgs = MemoryManager.malloc8d(nbins+1);
+      for( int i=0; i<nbins; i++ ) avgs[i] = hs.w(i)==0 ? -Double.MAX_VALUE /* value doesn't matter - see below for sending empty buckets (unseen levels) into the NA direction */: hs.wY(i) / hs.w(i); // Average response
+      avgs[nbins] = Double.MAX_VALUE;
+      ArrayUtils.sort(idxs, avgs);
+      // Fill with sorted data.  Makes a copy, so the original data remains in
+      // its original order.
+      vals = MemoryManager.malloc8d(vals_dim*nbins);
+      valsUplift = MemoryManager.malloc8d(4*nbins);
+
+      for( int i=0; i<nbins; i++ ) {
+        int id = idxs[i];
+        vals[vals_dim*i+0] = hs._vals[vals_dim*id+0];
+        vals[vals_dim*i+1] = hs._vals[vals_dim*id+1];
+        vals[vals_dim*i+2] = hs._vals[vals_dim*id+2];
+        valsUplift[valsUpliftDim * i] = valsUplift[valsUpliftDim * id];
+        valsUplift[valsUpliftDim * i + 1] = valsUplift[valsUpliftDim * id + 1];
+        valsUplift[valsUpliftDim * i + 2] = valsUplift[valsUpliftDim * id + 2];
+        valsUplift[valsUpliftDim * i + 3] = valsUplift[valsUpliftDim * id + 3];
+        if (LOG.isTraceEnabled()) LOG.trace(vals[3*i] + " obs have avg response [" + i + "]=" + avgs[id]);
+      }
+    }
+
+    // Compute mean/var for cumulative bins from 0 to nbins inclusive.
+    double   wlo[] = MemoryManager.malloc8d(nbins+1);
+    double  wYlo[] = MemoryManager.malloc8d(nbins+1);
+    double wYYlo[] = MemoryManager.malloc8d(nbins+1);
+    double[]  numloTreat = MemoryManager.malloc8d(nbins + 1);
+    double[]  resploTreat = MemoryManager.malloc8d(nbins + 1);
+    double[] numloContr = MemoryManager.malloc8d(nbins + 1);
+    double[]  resploContr = MemoryManager.malloc8d(nbins + 1);
+
+    for( int b = 1; b <= nbins; b++ ) {
+      int id = vals_dim * (b - 1);
```

Comment:
`b-1` could be stored in local variable since it is used 9-times

Reply:
Good point, but no refactor in this code is allowed within this PR.